File: 96.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution{
public:
	int numTrees(int n){
		vector<int> res(n+1,0);
		if(n>=1){
			res[0]=1;
			res[1]=1;
		}
		for(int i=2;i<=n;++i){
			for(int j=0;j<=i-1;++j){
				res[i]+=res[j]*res[i-1-j];
			}
		}
		return res[n];
	}
};

int main(){
	Solution s;
	int n=3;
	cout<<s.numTrees(n)<<endl;

	return 0;
}

File: Python/96.py
class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [0]*(n+1)
        dp[0] = 1
        for i in range(1, n+1):
            for j in range(i):
                dp[i] += dp[j]*dp[i-j-1]
        return dp[n]
